File: bbb/bbb/report/customer_wise_sales_summary/customer_wise_sales_summary.js
// Copyright (c) 2022, invento software limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Customer Wise Sales Summary"] = {
	"filters": [
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
		},
		{
			"fieldname":"pos_profile",
			"label": __("Outlet"),
			"fieldtype": "Link",
			"options": "POS Profile",
      "get_query": () => {
				var company = frappe.query_report.get_filter_value('company');
				return {
					filters: {
						'company': company
					}
				};
			},
      "on_change": function () {
        if(this.value === 'Distribution'){
          frappe.query_report.set_filter_value('switch_invoice', "Sales Invoice");
          frappe.query_report.refresh();
        }else{
          frappe.query_report.set_filter_value('switch_invoice', "POS Invoice");
          frappe.query_report.refresh();
        }
      }
		},
		{
			"fieldname":"switch_invoice",
			"label": __("Switch Invoice"),
			"fieldtype": "Data",
			"default": "POS Invoice",
			"reqd": 1,
      "read_only":1,
			"width": "60px"
		},
		{
			"fieldname":"customer",
			"label": __("Customer"),
			"fieldtype": "Link",
			"options": "Customer",
		},
    {
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
      "default": frappe.defaults.get_default('company'),
			"reqd": 1,
			"width": "60px"
		},
	]
};

File: bbb/bbb/report/customer_wise_sales_summary/customer_wise_sales_summary.py
# Copyright (c) 2022, invento software limited and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    columns, data = get_columns(), get_invoice_data(filters)
    return columns, data

def get_columns():
    """return columns"""
    columns = [
        {"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "POS Profile","width": 250},
        {"label": _("Mobile No."), "fieldname": "mobile_number", "fieldtype": "Data", "width": 190},
        # {"label": _("Customer Type"), "fieldname": "customer_type", "fieldtype": "Data", "width": 160},
        {"label": _("Customer Group"), "fieldname": "customer_group", "fieldtype": "Data", "width": 190},
        {"label": _("Outlet"), "fieldname": "pos_profile", "fieldtype": "Link", "options":"POS Profile", "width": 190},
        {"label": _("Total Invoice"), "fieldname": "total_invoice", "fieldtype": "Float", "width": 190},
        {"label": _("Total Amount"), "fieldname": "total_amount", "fieldtype": "Currency", "width": 190,
         "convertible": "rate", "options": "currency"},
    ]
    return columns


def get_conditions(filters):
    conditions = []

    if filters.get("from_date"):
        conditions.append("invoice.posting_date >= '%s'" % filters.get("from_date"))
    if filters.get("to_date"):
        conditions.append("invoice.posting_date <= '%s'" % filters.get("to_date"))
    if filters.get('pos_profile'):
        conditions.append(("invoice.pos_profile = '%s'" % filters.get("pos_profile")))
    if filters.get('customer'):
        conditions.append(("invoice.customer = '%s'" % filters.get("customer")))

    if conditions:
        conditions = " and ".join(conditions)
    return conditions


def get_invoice_data(filters):
    conditions = get_conditions(filters)
    invoice_type = filters.get('switch_invoice', "Sales Invoice")
    pos_profile = filters.get('pos_profile', '')

    invoice_query = """select count(invoice.name) as total_invoice, invoice.customer, sum(invoice.grand_total) as total_amount,
                        (case when '%s' = '' then 'All Outlet' else invoice.pos_profile end) as pos_profile
                        from `tab%s` invoice""" % (pos_profile, invoice_type)

    if conditions:
        invoice_query += """ where %s group by invoice.customer""" % conditions

    customer_sales_summary_query = """select total_invoice, total_amount, customer, customer.mobile_number, invoice_group.pos_profile,
                                    customer.customer_group, customer.customer_type from (%s) as invoice_group inner join `tabCustomer` customer 
                                    on customer.name = invoice_group.customer order by total_amount desc""" % invoice_query

    query_result = frappe.db.sql(customer_sales_summary_query, as_dict=1)
    # print(query_result)
    return query_result